### PR TITLE
Better generic unmanaged structs handling ++ bidirectional F#/C# interop for 'unmanaged' constraint

### PR DESCRIPTION
**Pull request by vzarytovskii**
Pretty much a continuation of @TIHan's work on on top of `main` (previous PR for reference: https://github.com/dotnet/fsharp/pull/6064).
RFC: https://github.com/fsharp/fslang-design/blob/main/RFCs/FS-1090-Generic-struct-type-whose-fields-are-all-unmanaged-types-is-unmanaged.md

**TODO/FIXES:**
- [x] Treat type arguments with IsUnmanagedAttribute as if they have the unmanaged constraint.
- [x] Emit `IsUnmanagedAttribute` for the `unmanaged` constraints on params.
- [x] Emit `UnmanagedType` modreq on type arguments on types 
       - This would be a binary breaking change, we should consider if we want to do it or not.
- [x] An `unmanaged` constraint cannot be used together with `not struct` constraint (not the case right now).
- [x] Fix struct tuples, and struct anonymous records - whenwe check `isUnmanaged` on their arguments, tuple/record hasn't been solved yet.

** TEST list ** (Tomas's braindump)
- F# side definition options:
   - [x] Generic custom types
   - [x] Generic struct records
   - [x] Struct tuples
   - [x] Struct DU's
   - [x] Struct anonymous records
   - [x] Deeply nested generic scenarios  ( `MyStruct<MyOtherStruct<StructTuple<int,byte>>>` )
   - [x] Any nested combination of the following (e.g. AnonRecord<StructTuple<CustomType<Record<int>>>>)
- Test matrix dimensions for interop (missing anything?)
    - [x]  Method with unmanaged constraint defined in C#/F#   (*2 test cases)
    - [x]  Data to be passed (a generic struct) defined in C#/F# (*2 test cases)
    - [x]  The T is either valid or invalid and should show error (*2 test cases)
    - [x] The callsite is C#/F# (*2 test cases)
- Additional test pointers:
-  If we do not put in a modreq in F# codegen for a constrained method,:
    - [x] C# can still call it
    - [x]  C# can still typecheck it for in/correct usage 
    - [ ]  VB.NET is not capable of calling it incorrectly
        - [ ] Well, apparently, VB.NET does not support this feature and is not capable of calling into such methods.
    - [x] For places where C# puts in a modreq, F# can call those methods (haven't checked deeply yet, might differ across virutal/non-vritual scenarios)

### CODE REVIEW

**Comment by vzarytovskii** [comment]
> Emit UnmanagedType modreq on type arguments.

I am not sure we should emit it just because C# does it as well, since it will change the default behaviour we already have, it will also be not easy to test.

**Comment by T-Gro** [comment]
Before this PR, the `unmanaged` constraint defined in F# was erased for IL output, therefore any other .NET consumer could call into such defined types/methods.

With this PR, the modreq is being emitted.
Which F# and C# understands, but VB.NET does not:

https://github.com/dotnet/vblang/issues/300 
https://github.com/dotnet/vblang/issues/604


Which means that public APIs that should also target VB.NET will need to be aware of this and provide alternatives.

**Comment by Happypig375** [comment]
`voidptr` support? https://github.com/fsharp/fslang-suggestions/issues/1160

**Comment by T-Gro** [comment]
> `voidptr` support? [fsharp/fslang-suggestions#1160](https://github.com/fsharp/fslang-suggestions/issues/1160)

I don't see why not, I can add it in a separate PR right after this one is merged in.

File: src/Compiler/FSComp.txt [hunk 1]
```diff
@@ -1709,4 +1709,5 @@ featureAccessorFunctionShorthand,"underscore dot shorthand for accessor only fun
...
 3570,tcStaticBindingInExtrinsicAugmentation,"Static bindings cannot be added to extrinsic augmentations. Consider using a 'static member' instead."
 3571,pickleFsharpCoreBackwardsCompatible,"Newly added pickle state cannot be used in FSharp.Core, since it must be working in older compilers+tooling as well. The time window is at least 3 years after feature introduction. Violation: %s . Context: \n %s "
-3577,tcOverrideUsesMultipleArgumentsInsteadOfTuple,"This override takes a tuple instead of multiple arguments. Try to add an additional layer of parentheses at the method definition (e.g. 'member _.Foo((x, y))'), or remove parentheses at the abstract method declaration (e.g. 'abstract member Foo: 'a * 'b -> 'c')."
\ No newline at end of file
+3577,tcOverrideUsesMultipleArgumentsInsteadOfTuple,"This override takes a tuple instead of multiple arguments. Try to add an additional layer of parentheses at the method definition (e.g. 'member _.Foo((x, y))'), or remove parentheses at the abstract method declaration (e.g. 'abstract member Foo: 'a * 'b -> 'c')."
+featureUnmanagedConstraintCsharpInterop,"Interop betwee C#'s and F#'s unmanaged generic constraint"
```
Comment (vzarytovskii) [inline src/Compiler/FSComp.txt:1713]:
```suggestion
featureUnmanagedConstraintCsharpInterop,"Interop between C#'s and F#'s unmanaged generic constraint (emit additional modreq)"
```

**Review by 0101 — APPROVED** [review]

**Pull request merged**